target/linux/ath79/base-files/lib/upgrade/platform.sh

```diff
@@ -5,14 +5,112 @@
 PART_NAME=firmware
 REQUIRE_IMAGE_METADATA=1
 
-platform_check_image() {
+CI_BLKSZ=65536
+CI_LDADR=0x80060000
+
+PLATFORM_DO_UPGRADE_COMBINED_SEPARATE_MTD=0
+
+platform_find_partitions() {
+	local first dev size erasesize name
+	while read dev size erasesize name; do
+		name=${name#'"'}; name=${name%'"'}
+		case "$name" in
+			vmlinux.bin.l7|vmlinux|kernel|linux|linux.bin|rootfs|filesystem)
+				if [ -z "$first" ]; then
+					first="$name"
+				else
+					echo "$erasesize:$first:$name"
+					break
+				fi
+			;;
+		esac
+	done < /proc/mtd
+}
+
+platform_find_kernelpart() {
+	local part
+	for part in "${1%:*}" "${1#*:}"; do
+		case "$part" in
+			vmlinux.bin.l7|vmlinux|kernel|linux|linux.bin)
+				echo "$part"
+				break
+			;;
+		esac
+	done
+}
+
+platform_find_rootfspart() {
+	local part
+	for part in "${1%:*}" "${1#*:}"; do
+		[ "$part" != "$2" ] && echo "$part" && break
+	done
+}
+
+platform_check_combined() {
+	[ "$magic" != "4349" ] && {
+		echo "Invalid image. Use proper *-sysupgrade.bin file for this board"
+		return 1
+	}
+
 	return 0
 }
 
+platform_do_upgrade_combined() {
```

Comment:
I don't really like the combined image format. In my opinion it is way to complex and depends on a lot of magic. For the [OpenMesh A62](0b204902075157dbb002ea04f41a5b1a4fbe848c) we managed to reuse parts of the nand upgrade code to write to the -not in a row - kernel and rootfs partition.

I'm not yet sure if we can do the same here, as I would expect that you would like to have a migration path from the ar71xx image to the ath79 image.

Before going into more details I would like your opinion about a new/different sysupgrade format.

Reply:
Thanks for providing another in-house solution, didn't know about this one. I'll have to look closely at this and experiment a bit, at first glance it looks pretty clean.

Personally don't have strong feelings about `combined-image`, I just went with old code which worked. Whichever solution is underneath doesn't matter for end user. What matters is swift upgrade, that's why I would like to keep the possibility to sysupgrade from ar71xx. Right now I can't say when I'll try to accommodate it, so I would like to keep combined-image for now.